### PR TITLE
Merl 1153 faustwp block editor block controls implement checkbox control

### PR DESCRIPTION
**Pull request by TeresaGobble**
## Tasks

- [x] I have signed a [Contributor License Agreement (CLA)](https://github.com/wpengine/faustjs#contributor-license-agreement) with WP Engine.
- [x] If a code change, I have written testing instructions that the whole team & outside contributors can understand.
- [x] I have written and included a comprehensive [changeset](https://github.com/wpengine/faustjs/blob/canary/DEVELOPMENT.md#deployment) to properly document the changes I've made.

## Description

This PR implements the checkbox control feature for the `block-editor-utils` package. Its corresponding jira ticket is [here](https://wpengine.atlassian.net/jira/software/c/projects/MERL/boards/1007?modal=detail&selectedIssue=MERL-1153).

## Testing

After pulling this branch down, run the following commands in the faustjs monorepo:
1. run `npm install` 
2. run `npm run build` 
3. run `npm run build -w packages/block-editor-utils`
4. run `cd packages/block-editor-utils`
5. run `npm link`

Switch to your testing site's repo.

6. On your WordPress instance create a new block plugin using [@wordpress/create-block](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-create-block/)
npx @wordpress/create-block hello-world-block
7. cd to your new block plugin's folder 
8. Inside the `hello-world-block` plugin folder, run `npm link @faustwp/block-editor-utils` to add a reference to the package.
9. Replace `hello-world-block`'s generated index.js file with the below code block:
```
// In app/public/wp-content/plugins/hello-world-block/src/index.js:

import { addFilter } from '@wordpress/hooks';
import "./style.scss";

// import block.json
import metadata from "./block.json";

// import Pure React Component
import HelloWorld from "./HelloWorld";

// Register React Component in Gutenberg
import { registerFaustBlock } from "@faustwp/block-editor-utils";

registerFaustBlock(HelloWorld, { blockJson: metadata });
```
10. Replace `hello-world-block`'s generated `block.json file with the below code block.

In `app/public/wp-content/plugins/hello-world-block/src/block.json`:
```
{
	"$schema": "https://schemas.wp.org/trunk/block.json",
	"apiVersion": 2,
	"name": "create-block/hello-world-block",
	"version": "0.1.0",
	"title": "Hello World Block",
	"category": "widgets",
	"icon": "smiley",
	"description": "Example block scaffolded with Create Block tool.",
	"supports": {
		"html": false
	},
	"attributes": {
		"message": {
			"type": "string",
			"default": "Hello World"
		},
		"checkbox": {
			"type": "boolean",
			"default": ""
		}
	},
	"textdomain": "hello-world-block",
	"editorScript": "file:./index.js",
	"editorStyle": "file:./index.css",
	"style": "file:./style-index.css"
}
```
11. Add a file named `HelloWorld.js` and add the below code block.
```
// In app/public/wp-content/plugins/hello-world-block/src/HelloWorld.js:

function HelloWorld({ style, className, attributes, children, ...props }) {
	const styles = {
		...style,
		backgroundColor: attributes.bg_color,
		color: attributes.text_color,
	};
	return (
		<div
			{...props}
			style={styles}
			className={className}
			dangerouslySetInnerHTML={{ __html: attributes.message }}
		/>
	);
}

HelloWorld.config = {
	name: "HelloWorld",
        editorFields: {
		message: {
			type: "string",
			default: "",
			label: "My Message",
			location: "editor",
		},
		checkbox: {
			location: "inspector",
			control: "checkbox",
			label: "My Checkbox",
			default: "false",
			help: "here is a help message"
		},
	},
};
export default HelloWorld;
```
12. Run `npm start` from your plugin's file, `hello-world-block`.
13. Ensure your block plugin, `Hello World Block` is active.
14. Go into the editor and try to create a new `Hello World Block` block. You should look to the sidebar and see the radioField panel visible with whatever options you pass into it. Ensure that upon selecting a value that it persists between saves and edits.

## Screenshots

image

### CODE REVIEW

**Comment by TeresaGobble** [comment]
I have a few open questions on this PR requirement, @theodesp and @matthewguywright:

1. The config I've created has a `help` [attribute type](https://developer.wordpress.org/block-editor/reference-guides/components/checkbox-control/#help-stringwpelement) as optional in the `packages/block-editor-utils/src/types/index.d.ts` file. I notice where the `help` attribute is omitted in [a similar PR](https://github.com/wpengine/faustjs/pull/1571/files#r1323581798) and am wondering whether this is intentional. I have added it to the control, but it does not appear to be showing up in the editor. 
2. Do all the controls work recursively? I want to ensure that all checkbox children of a parent checkbox are marked or unmarked, [depending on the state of the parent checkbox](https://developer.wordpress.org/block-editor/reference-guides/components/checkbox-control/#parent-and-child-checkboxes). I'm not sure how I can check to ensure this is working correctly- I've already tried a few variations of nesting checkboxes in the `block.json` and my test plugin, but it isn't showing up in the editor as expected.
3. The testing I have written uses Matt's example from the RadioControl component. It passes locally, but I'd like to take another look with someone to ensure I understand the logic behind it, in particular the value set by the `props` and how the test then checks for a truthy value.

Thank you for your time in reviewing this item with me, and please let me know if I can clarify anything on my questions above! I appreciate your assistence. 😊

File: packages/block-editor-utils/src/controls/Checkbox.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+}: ControlProps<T>) {
+  const onChange = (newContent: boolean) => {
+    props.setAttributes({ [config.name]: newContent });
+  };
+  return (
+    // TODO: ensure contents match CheckControl in Block Editor- what is help value?
+    // https://developer.wordpress.org/block-editor/reference-guides/components/checkbox-control/#usage-2
+    <CheckboxControl
+      label={config.label}
+      help={props.attributes[config.help]}
```
Comment (theodesp) [inline packages/block-editor-utils/src/controls/Checkbox.tsx:17]: We cannot assume a good help label for now. Let's remove this.

File: packages/block-editor-utils/src/types/index.d.ts [hunk 2]
```diff
@@ -19,6 +19,7 @@ export type Field = {
   location: FieldLocation;
   label?: string;
   default?: unknown;
+  help?: string;
```
Comment (theodesp) [inline packages/block-editor-utils/src/types/index.d.ts:22]: Let's remove this since we don't really need help labels in our controls.

**Comment by theodesp** [comment]
> I have a few open questions on this PR requirement, @theodesp and @matthewguywright:
> 
> 1. The config I've created has a `help` [attribute type](https://developer.wordpress.org/block-editor/reference-guides/components/checkbox-control/#help-stringwpelement) as optional in the `packages/block-editor-utils/src/types/index.d.ts` file. I notice where the `help` attribute is omitted in [a similar PR](https://github.com/wpengine/faustjs/pull/1571/files#r1323581798) and am wondering whether this is intentional. I have added it to the control, but it does not appear to be showing up in the editor.
> 2. Do all the controls work recursively? I want to ensure that all checkbox children of a parent checkbox are marked or unmarked, [depending on the state of the parent checkbox](https://developer.wordpress.org/block-editor/reference-guides/components/checkbox-control/#parent-and-child-checkboxes). I'm not sure how I can check to ensure this is working correctly- I've already tried a few variations of nesting checkboxes in the `block.json` and my test plugin, but it isn't showing up in the editor as expected.
> 3. The testing I have written uses Matt's example from the RadioControl component. It passes locally, but I'd like to take another look with someone to ensure I understand the logic behind it, in particular the value set by the `props` and how the test then checks for a truthy value.
> 
> Thank you for your time in reviewing this item with me, and please let me know if I can clarify anything on my questions above! I appreciate your assistence. 😊

Hey @TeresaGobble. Let's get into them:

1. I would rather leave the help label out since it doesn't serve any purpose and not all controls have them.
2. You don't need to worry about recursive blocks or parent references. We will only support a single checkbox control that is controlled via a single block.json attribute for now to keep it simple.
3. You can expand the. testing cases by simulating click events. Take a look at the [api-events](https://testing-library.com/docs/dom-testing-library/api-events/) documentation for example.

File: packages/block-editor-utils/src/controls/Checkbox.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+function Checkbox<T extends Record<string, any>>({
+  config,
+  props,
+}: ControlProps<T>) {
+  const onChange = (newContent: boolean) => {
+    props.setAttributes({ [config.name]: newContent });
+  };
+  return (
+    // TODO: ensure contents match CheckControl in Block Editor- what is help value?
```
Comment (matthewguywright) [inline packages/block-editor-utils/src/controls/Checkbox.tsx:13]: Based on @theodesp answers, you can remove this TODO.

File: packages/block-editor-utils/tests/controls/Checkbox.test.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    input,
+    ...utils,
+  };
+};
+
+afterEach(() => {
+  jest.clearAllMocks();
+});
+
+describe('Checkbox', () => {
```
Comment (matthewguywright) [inline packages/block-editor-utils/tests/controls/Checkbox.test.tsx:52]: Nice test!

File: packages/block-editor-utils/src/controls/Checkbox.tsx (hunk 3 above)
Comment (TeresaGobble) [reply packages/block-editor-utils/src/controls/Checkbox.tsx:13]: Thank you, will do!

**Review by matthewguywright — APPROVED** [review]

**Pull request merged**